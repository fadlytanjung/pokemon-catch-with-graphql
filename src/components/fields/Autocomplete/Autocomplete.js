import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import styles from './styles.scoped.css';

export default function Autocomplete(props) {
  const { className, input, inputProps, meta, onSearch, options } = props;
  const { error } = meta;
  const isInitialMount = useRef(true);
  const [search, setSearch] = useState('');
  const [timer, setTimer] = useState(0);
  const [value, setValue] = useState('');
  const [display, setDisplay] = useState('none');
  const isError = !!error && !inputProps.disabled;
  const classes = [
    styles.root,
    isError && styles.error,
    className,
  ].filter(Boolean).join(' ');

  const onClickItem = item => () => {
    setSearch(item.text);
    setValue(item.value);
    setDisplay('none');
  };

  const onBlur = () => {
    (!options.length) && setDisplay('none');
  };

  const onChange = e => {
    const targetValue = e.target.value;
    setSearch(targetValue);
    setValue('');
    setDisplay('block');
  };

  const onFocus = () => {
    setDisplay('block');
  };

  const inputEvents = { onBlur, onChange, onFocus };

  useEffect(() => {
    (!input.value) && setSearch('');
  }, [input.value]);

  useEffect(() => {
    input.onChange(value);
  }, [value]);

  useEffect(() => {
    if (isInitialMount.current) {
      isInitialMount.current = false;
    } else {
      clearTimeout(timer);
      const timeout = setTimeout(() => {
        onSearch(search);
      }, 500);
      setTimer(timeout);
    }
  }, [search]);

  return (
    <>
      <label className={styles.label}>{inputProps.label}</label>
      <input type="hidden" {...input} />
      <input className={classes} value={search} {...inputEvents} {...inputProps} />
      <div className={styles.options}>
        <ul style={{ display }}>
          {options.map((item, key) => (
            <li key={key} onClick={onClickItem(item)}>
              <ItemText search={search} text={item.text} />
            </li>
          ))}
        </ul>
      </div>
      {isError && <small className={styles.error}>{error}</small>}
    </>
  );
}

Autocomplete.defaultProps = {
  className: '',
  input: { onChange: () => {} },
  inputProps: {},
  meta: {},
  onSearch: () => { },
  options: [],
};

Autocomplete.propTypes = {
  className: PropTypes.string,
  input: PropTypes.object,
  inputProps: PropTypes.object,
  meta: PropTypes.object,
  onSearch: PropTypes.func,
  options: PropTypes.arrayOf(PropTypes.shape({
    text: PropTypes.string,
    value: PropTypes.string,
  })),
};

export function ItemText(props) {
  const { search, text } = props;
  const match = text.match(new RegExp(search.replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1'), 'i'));
  if (!match) {
    return text;
  }
  return (
    <>
      {text.substring(0, match.index)}
      {<span>{text.substring(match.index, match.index + search.length)}</span>}
      {text.substring(match.index + search.length)}
    </>
  );
}

ItemText.defaultProps = {
  search: '',
  text: '',
};

ItemText.propTypes = {
  search: PropTypes.string,
  text: PropTypes.string,
};
