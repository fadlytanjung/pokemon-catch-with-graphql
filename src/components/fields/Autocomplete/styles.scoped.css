.root {
  background: url(/assets/ic_chevron_red_down.svg);
  background-color: #FFFFFF;
  background-position: calc(100% - 0.75rem);
  background-repeat: no-repeat;
  border: 0.0625rem solid #cccccc;
  border-radius: 0.5rem;
  box-shadow: none;
  color: var(--light-text-color);
  display: block;
  font-size: 0.9rem;
  font-weight: 400;
  outline: none;
  padding: 0.75rem 3rem 0.75rem 0.75rem;
  transition: border-color 0.15s ease-in-out, border-radius 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
  width: 100%;
}

.root:not(.error):focus {
  border-color: #808080;
  box-shadow: 0 0 0 0.0625rem #808080;
  color: #333435;
}

.root:disabled {
  background-color: #e5e5e5 !important;
  color: #999999 !important;
}

.root::placeholder {
  color: #b2b4b5;
}

.root.error {
  border-color: #ee4c24;
}

.root:focus {
  border-radius: 0.5rem 0.5rem 0 0;
}

.options {
  position: relative;
  z-index: 90;
}

.options > ul {
  background-color: #ffffff;
  border: 0.0625rem solid #cccccc;
  border-radius: 0 0 0.5rem 0.5rem;
  box-shadow: 0 0 0 0.0625rem #808080;
  max-height: 10rem;
  overflow-y: auto;
  padding: 0 0.75rem;
  position: absolute;
  top: 0;
  width: 100%;
}

.options > ul > li {
  border-bottom: 0.0625rem solid #cccccc;
  cursor: pointer;
  font-size: 0.9rem;
  padding: 0.75rem 0;
}

.options > ul > li > span {
  animation: none;
  font-weight: 800;
}

.options > ul > li:hover, .options > ul > li:hover > span {
  color: var(--primary-color);
}

.options + small.error {
  color: #ee4c24;
  display: block;
  font-size: 0.75rem;
  margin-top: 0.25rem;
}

.label {
  display: inline-block;
  margin-bottom: 0.5rem;
  text-transform: uppercase;
  letter-spacing: 0.0625rem;
  color: #999999;
  font-weight: 400;
  font-size: 0.75rem;
}
