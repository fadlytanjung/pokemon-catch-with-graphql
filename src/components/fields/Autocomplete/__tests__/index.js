import React, { useRef, useState } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Autocomplete, { ItemText } from '../Autocomplete';

describe('src/components/fields/Autocomplete', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Autocomplete options={[{ text: '', value: '' }]} />);
    Autocomplete.defaultProps.input.onChange();
    Autocomplete.defaultProps.onSearch();
    expect(tree).toMatchSnapshot();
  });

  test('error message & timer', () => {
    const props = {
      ...Autocomplete.defaultProps,
      meta: { error: 'error' },
      onSearch: jest.fn(),
    };
    global.clearTimeout = jest.fn();
    global.setTimeout = fn => fn();
    const setTimer = jest.fn();
    useState.mockImplementationOnce(() => ['search', jest.fn()])
      .mockImplementationOnce(v => [v, setTimer]);
    useRef.mockImplementationOnce(() => ({ current: false }));
    const result = Autocomplete(props);
    expect(result.props.children[4].props.children).toBe('error');
    expect(props.onSearch).toHaveBeenCalledWith('search');
    expect(setTimer).toHaveBeenCalled();
  });

  test('onClickItem, onBlur, onChange, & onFocus', () => {
    const props = {
      ...Autocomplete.defaultProps,
      input: { onChange: jest.fn(), value: 'inputValue' },
      options: [{ text: 'text', value: 'value' }],
    };
    const setSearch = jest.fn();
    const setValue = jest.fn();
    const setDisplay = jest.fn();
    useState.mockImplementationOnce(v => [v, setSearch])
      .mockImplementationOnce(v => [v, jest.fn()])
      .mockImplementationOnce(v => [v, setValue])
      .mockImplementationOnce(v => [v, setDisplay]);
    const result = Autocomplete(props);
    result.props.children[3].props.children.props.children[0].props.onClick();
    expect(setSearch).toHaveBeenNthCalledWith(1, 'text');
    expect(setValue).toHaveBeenNthCalledWith(1, 'value');
    expect(setDisplay).toHaveBeenNthCalledWith(1, 'none');

    result.props.children[2].props.onBlur();
    result.props.children[2].props.onChange({ target: { value: 'targetValue' } });
    expect(setSearch).toHaveBeenNthCalledWith(2, 'targetValue');
    expect(setValue).toHaveBeenNthCalledWith(2, '');
    expect(setDisplay).toHaveBeenNthCalledWith(2, 'block');

    result.props.children[2].props.onFocus();
    expect(setDisplay).toHaveBeenNthCalledWith(3, 'block');

    useState.mockImplementationOnce(v => [v, setSearch])
      .mockImplementationOnce(v => [v, jest.fn()])
      .mockImplementationOnce(v => [v, setValue])
      .mockImplementationOnce(v => [v, setDisplay]);
    const result2 = Autocomplete({ ...props, options: [], });
    result2.props.children[2].props.onBlur();
    expect(setDisplay).toHaveBeenNthCalledWith(4, 'none');
  });

  test('ItemText', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<ItemText />);
    expect(tree).toMatchSnapshot();

    const result = ItemText({ search: 'tes', text: 'telkom' });
    expect(result).toBe('telkom');
  });
});
