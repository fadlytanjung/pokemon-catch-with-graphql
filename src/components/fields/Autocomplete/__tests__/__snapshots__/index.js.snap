// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/fields/Autocomplete ItemText 1`] = `
<React.Fragment>
  
  <span>
    
  </span>
  
</React.Fragment>
`;

exports[`src/components/fields/Autocomplete render 1`] = `
<React.Fragment>
  <label />
  <input
    onChange={[Function]}
    type="hidden"
  />
  <input
    className=""
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value=""
  />
  <div>
    <ul
      style={
        Object {
          "display": "none",
        }
      }
    >
      <li
        onClick={[Function]}
      >
        <ItemText
          search=""
          text=""
        />
      </li>
    </ul>
  </div>
</React.Fragment>
`;
